cpp/include/kvikio/remote_handle.hpp

```diff
@@ -0,0 +1,356 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cstdlib>
+#include <iostream>
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <utility>
+
+#include <aws/core/Aws.h>
+#include <aws/core/auth/AWSCredentialsProviderChain.h>
+#include <aws/core/utils/stream/PreallocatedStreamBuf.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/GetObjectRequest.h>
+#include <aws/s3/model/HeadObjectRequest.h>
+
+#include <kvikio/parallel_operation.hpp>
+#include <kvikio/posix_io.hpp>
+#include <kvikio/utils.hpp>
+
+#ifndef KVIKIO_AWS_SDK_FOUND
+#error "cannot include <kvikio/remote_handle.hpp>, configuration did not find AWS SDK"
+#endif
+
+namespace kvikio {
+namespace detail {
+
+/**
+ * Stream implementation of a fixed size buffer.
+ */
+class BufferAsStream : public Aws::IOStream {
+ public:
+  using Base = Aws::IOStream;
+  explicit BufferAsStream(std::streambuf* buf) : Base(buf) {}
+
+  ~BufferAsStream() override = default;
+};
+
+/**
+ * @brief Given a file path like "s3://<bucket>/<object>", return the name of the bucket and object.
+ *
+ * @param path S3 file path.
+ * @return Pair of strings: [bucket-name, object-name].
```

Comment:
*nit*: Document the handling for `s3://bucket` and `s3://bucket/` As well as `@throws std::invalid_argument`